t/01-distributed.t

```diff
@@ -0,0 +1,205 @@
+    $history = $dbh1_2->{mock_all_history};
+    is(scalar @$history, 3, "3 statement");
+    is($history->[0]->statement, 'BEGIN WORK');
+    is($history->[1]->statement, 'select 1_2');
+    is($history->[2]->statement, 'COMMIT');
+    $history = $dbh2_1->{mock_all_history};
+    is(scalar @$history,         1,            "dbh2_1 is not in category1, so only one statement");
+    is($history->[0]->statement, 'select 2_1', 'dbh2_1 no begin_work and commit');
+
+    for my $dbh ($dbh1_1, $dbh1_2, $dbh2_1) {
+        $dbh->{mock_clear_history} = 1;
+        $dbh->{AutoCommit}         = 1;    # DBD::Mock has an bug. the second dbh cannot reset autocommit after commit. so we reset it by hand
+    }
+
+    my @result = txn(sub { $code->($dbh1_1, $dbh1_2, $dbh2_1) }, 'category1');
+    is_deeply(\@result, [qw(1_1 1_2 2_1)], 'wantarray will get a list');
+
+    for my $dbh ($dbh1_1, $dbh1_2, $dbh2_1) {
+        $dbh->{mock_clear_history} = 1;
+        $dbh->{AutoCommit}         = 1;    # DBD::Mock has an bug. the second dbh cannot reset autocommit after commit. so we reset it by hand
+    }
+
+    $dbh1_2->{mock_add_parser} = sub {
+        my $sql = shift;
+        die "simulate parse error"    if $sql =~ /select/;
+        die "simulate rollback error" if $sql =~ /rollback/i;
+    };
+    warnings_like {
+        throws_ok {
+            txn(sub { $code->($dbh1_1, $dbh1_2, $dbh2_1) }, 'category1');
+        }
+        qr/simulate parse error/, "will die if error";
+    }
+    [qr/simulate parse error/, qr/Error in transaction/, qr/simulate rollback error/, qr/^after/is], "have warnings";
+    $history = $dbh1_1->{mock_all_history};
+    is($history->[-1]->statement, 'ROLLBACK', 'dbh1_1 rollbacked');
+
+    # test fork
+    clear_dbh_for_txn_test();
+    ($dbh1_1, $dbh1_2, $dbh2_1) = init_dbh_for_txn_test();
+    local $$ = fake_pid();    # change $$ to simulate fork
+    lives_ok {
+        txn(sub { $code->($dbh1_1, $dbh1_2, $dbh2_1) }, 'category1');
+    };
+    $history = $dbh1_1->{mock_all_history};
+    is(scalar @$history,          1,            'no begin_work and rollback because fork will clear all registered dbh');
+    is($history->[-1]->statement, 'select 1_1', 'only select statement');
+
+    clear_dbh_for_txn_test();
+    ($dbh1_1, $dbh1_2, $dbh2_1) = init_dbh_for_txn_test();
+    $code = sub {
+        my ($dbh1_1, $dbh1_2, $dbh_2_1) = @_;
+        $$ = fake_pid();      # change $$ to simulate fork
+        $dbh1_1->do('select 1_1');
+        $dbh1_2->do('select 1_2');
+        $dbh2_1->do('select 2_1');
+        return wantarray ? qw(1_1 1_2 2_1) : "1";
+    };
+    lives_ok {
+        txn(sub { $code->($dbh1_1, $dbh1_2, $dbh2_1) }, 'category1');
+    };
+    $history = $dbh1_1->{mock_all_history};
+    is(scalar @$history,         2,            'no commit because fork will clear all registered dbh');
+    is($history->[0]->statement, 'BEGIN WORK', 'has begin work');
+    is($history->[1]->statement, 'select 1_1', 'has select');
+
```

Comment:
@binarytom I have uploaded this module to PAUSE just now . Can we fix them in the next version ?